### ISSUE TITLE
maintaining scroll position when loading older messages

### ISSUE THREAD

**Issue by vshahpledge** [jessesquires/JSQMessagesViewController#1838]
I've recently implemented paging in my chat view. I'm currently in the process of trying to get an infinite scroll feature implemented to go along with paging. The issue I'm having is with maintaining scroll position when loading new messages at the top of the collectionview. When I load my messages and call reloadData on the collectionview the collectionview's contentOffset needs to be adjusted. It's easy enough to do the following:

    let oldContentOffsetFromBottomY = collectionView.contentSize.height - collectionView.contentOffset.y

    dataSource.loadMore()
    collectionView.reloadData()
    collectionView.layoutIfNeeded()

    let newContentOffset = CGPointMake(collectionView.contentOffset.x, collectionView.contentSize.height - oldContentOffsetFromBottomY)
    collectionView.setContentOffset(newContentOffset, animated: false)

but setting the contentOffset in this way makes the collectionview stop scrolling (in the scenario that the user flicked their finger to make the chat scroll quickly). I'm trying to go for a Facebook Messenger type infinite scroll. Therefore-

How can I adjust the contentOffset of my collectionview to maintain scroll position and preserve scrolling velocity after adding more cells to the top of the collectionview?

**Comment by vshahpledge** [comment]
This is how I ended up achieving this:

    class WPMessagesCollectionViewFlowLayout: JSQMessagesCollectionViewFlowLayout {
        private var oldContentSizeHeight: CGFloat = 0.0
        private var contentSizeDelta: CGFloat?
        
        override func invalidateLayoutWithContext(context: UICollectionViewLayoutInvalidationContext) {
            if let contentSizeDelta = contentSizeDelta {
                context.contentOffsetAdjustment = CGPointMake(context.contentOffsetAdjustment.x, contentSizeDelta)
            }
            contentSizeDelta = nil
            
            super.invalidateLayoutWithContext(context)
        }
        
        override func collectionViewContentSize() -> CGSize {
            let newContentSize = super.collectionViewContentSize(),
            newContentSizeHeight = newContentSize.height
            
            if newContentSizeHeight != oldContentSizeHeight {
                contentSizeDelta = newContentSizeHeight - oldContentSizeHeight
            }
            oldContentSizeHeight = newContentSizeHeight
            
            return newContentSize
        }
    }

Basically, I take the old content size of the collectionview before it changes, the new content size after it changes, and use those two values to calculate a delta. I then use that delta to construct a CGPoint for the contentOffsetAdjustment property of the context in invalidateLayoutWithContext:

**Comment by gmansata6290** [comment]
@vshahpledge I just wanted ask you that If I create Subclass of JSQMessagesCollectionViewFlowLayout then all the default logic to handle the size of bubble again we have to write in our own class. Because I tried your lmplementation in my project but the bubble size calculation is getting messed up. So do you have a wayout where I can use the above implementation to existing JSQMessagesCollectionViewFlowLayout class only without disturbing the existing implmentation?

**Comment by gmansata6290** [comment]
Sorry my bad, wrong question I asked!! Thanks for the solution @vshahpledge

**Comment by gmansata6290** [comment]
@vshahpledge Thank you for your suggestion above. It worked well for maintaining the scroll position after loading more messages. But there is a small issue I am facing, (only for the first time) If my local db doesn't have anything then it calls an api, get the data and store it in db, again fetches from the db and load the messages. So at this moment collectionView content offset is going beyond the frame. So no messages showing, but when I scroll down, my messages are showing properly. And If I am loading messages from db without calling api then it works fine. 

Did you face this issue as well and if not, anything you could point me to so I can debug this?
